Provider.orElse(<provider>) tests presence eagerly

<!--- 
Please follow the instructions below. 
We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- Tell us what should happen -->
If I have a provider used in a task, it should not test presence during configuration

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
It does, and results in an exception

```
Caused by: java.lang.IllegalStateException: Cannot obtain value from provider of environment variable 'USER' at configuration time.
Use a provider returned by 'forUseAtConfigurationTime()' instead.
```

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
I want to use an environmentVariable provider and want to default it if absent.

### Steps to Reproduce 
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->

Drop this into a build.gradle.kts file

```kotlin
val userProvider = project.providers.environmentVariable("USER")
    .orElse(project.provider { "absent" })

abstract class TestTask : DefaultTask() {
  @get:Input
  abstract val userProp: Property<String>

  @TaskAction
  fun printUser() {
    println(userProp.get())
  }
}

project.tasks.register<TestTask>("testTask") {
  userProp.set(userProvider)
}
```

Then run `./gradlew testTask`

You'll see this failure

```
Could not determine the dependencies of task ':testTask'.
> Cannot obtain value from provider of environment variable 'USER' at configuration time.
  Use a provider returned by 'forUseAtConfigurationTime()' instead.

```

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildOptionsIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildOptionsIntegrationTest.groovy
@@ -19,7 +19,95 @@ package org.gradle.configurationcache
 import spock.lang.Issue
 import spock.lang.Unroll
 
+import static org.junit.Assume.assumeFalse
+
 class ConfigurationCacheBuildOptionsIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    @Issue("https://github.com/gradle/gradle/issues/13333")
+    @Unroll
+    def "absent #operator orElse #orElseKind used as task input"() {
+
+        assumeFalse(
+            'task dependency inference for orElse(taskOutput) not implemented yet!',
+            orElseKind == 'task output'
+        )
+
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+        buildKotlinFile """
+            abstract class PrintString : DefaultTask() {
+                @get:Input
+                abstract val string: Property<String>
+                @TaskAction
+                fun printString() {
+                    println("The string is " + string.get())
+                }
+            }
+            abstract class ProduceString : DefaultTask() {
+                @get:OutputFile
+                abstract val outputFile: RegularFileProperty
+                @TaskAction
+                fun printString() {
+                    outputFile.get().asFile.writeText("absent")
+                }
+            }
+            val producer = tasks.register<ProduceString>("produceString") {
+                outputFile.set(layout.buildDirectory.file("output.txt"))
+            }
+            val stringProvider = providers
+                .$operator("string")
+                .orElse($orElseArgument)
+            tasks.register<PrintString>("printString") {
+                string.set(stringProvider)
+            }
+        """
+        def printString = { string ->
+            switch (operator) {
+                case 'systemProperty':
+                    configurationCacheRun "printString", "-Dstring=$string"
+                    break
+                case 'gradleProperty':
+                    configurationCacheRun "printString", "-Pstring=$string"
+                    break
+                case 'environmentVariable':
+                    withEnvironmentVars(string: string)
+                    configurationCacheRun "printString"
+                    break
+            }
+        }
+
+        when:
+        configurationCacheRun "printString"
+
+        then:
+        output.count("The string is absent") == 1
+        configurationCache.assertStateStored()
+
+        when:
+        printString "alice"
+
+        then:
+        output.count("The string is alice") == 1
+        configurationCache.assertStateLoaded()
+
+        when:
+        printString "bob"
+
+        then:
+        output.count("The string is bob") == 1
+        configurationCache.assertStateLoaded()
+
+        where:
+        [operator, orElseKind] << [
+            ['systemProperty', 'gradleProperty', 'environmentVariable'],
+            ['primitive', 'provider', 'task output']
+        ].combinations()
+        orElseArgument = orElseKind == 'primitive'
+            ? '"absent"'
+            : orElseKind == 'provider'
+            ? 'providers.provider { "absent" }'
+            : 'producer.flatMap { it.outputFile }.map { it.asFile.readText() }'
+    }
 
     @Unroll
     def "system property from #systemPropertySource used as task and build logic input"() {

--- a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java
+++ b/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java
@@ -347,11 +347,6 @@ public abstract class AbstractFileCollection implements FileCollectionInternal {
         public ValueProducer getProducer() {
             return new ValueProducer() {
                 @Override
-                public boolean isKnown() {
-                    return true;
-                }
-
-                @Override
                 public boolean isProducesDifferentValueOverTime() {
                     return false;
                 }

--- a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/DefaultValueSourceProviderFactory.java
+++ b/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/DefaultValueSourceProviderFactory.java
@@ -21,6 +21,7 @@ import org.gradle.api.Describable;
 import org.gradle.api.GradleException;
 import org.gradle.api.NonExtensible;
 import org.gradle.api.internal.properties.GradleProperties;
+import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.provider.ValueSource;
 import org.gradle.api.provider.ValueSourceParameters;
@@ -260,12 +261,11 @@ public class DefaultValueSourceProviderFactory implements ValueSourceProviderFac
 
         @Override
         public ValueProducer getProducer() {
-            // For now, assume value is never calculated from a task output
-            if (value.hasBeenObtained()) {
-                return ValueProducer.unknown();
-            } else {
-                return ValueProducer.externalValue();
-            }
+            return ValueProducer.externalValue();
+        }
+
+        @Override
+        public void visitDependencies(TaskDependencyResolveContext context) {
         }
 
         @Override

--- a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/OrElseProvider.java
+++ b/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/OrElseProvider.java
@@ -16,7 +16,11 @@
 
 package org.gradle.api.internal.provider;
 
+import org.gradle.api.Action;
+import org.gradle.api.Task;
+
 import javax.annotation.Nullable;
+import java.util.ArrayList;
 
 class OrElseProvider<T> extends AbstractMinimalProvider<T> {
     private final ProviderInternal<T> left;
@@ -35,11 +39,10 @@ class OrElseProvider<T> extends AbstractMinimalProvider<T> {
 
     @Override
     public ValueProducer getProducer() {
-        if (left.isPresent()) {
-            return left.getProducer();
-        } else {
-            return right.getProducer();
-        }
+        ExecutionTimeValue<? extends T> leftValue = left.calculateExecutionTimeValue();
+        return leftValue.isMissing()
+            ? right.getProducer()
+            : new OrElseValueProducer();
     }
 
     @Override
@@ -69,5 +72,45 @@ class OrElseProvider<T> extends AbstractMinimalProvider<T> {
             return rightValue;
         }
         return leftValue.addPathsFrom(rightValue);
+    }
+
+    private class OrElseValueProducer implements ValueProducer {
+
+        final ValueProducer leftProducer = left.getProducer();
+        final ValueProducer rightProducer = right.getProducer();
+
+        @Override
+        public boolean isKnown() {
+            return leftProducer.isKnown()
+                || rightProducer.isKnown();
+        }
+
+        @Override
+        public boolean isProducesDifferentValueOverTime() {
+            return leftProducer.isProducesDifferentValueOverTime()
+                || rightProducer.isProducesDifferentValueOverTime();
+        }
+
+        @Override
+        public void visitProducerTasks(Action<? super Task> visitor) {
+            ArrayList<Task> leftTasks = producerTasksOf(leftProducer);
+            ArrayList<Task> rightTasks = producerTasksOf(rightProducer);
+            if (leftTasks.isEmpty() && rightTasks.isEmpty()) {
+                return;
+            }
+            // TODO: configuration cache: this condition needs to be evaluated at execution time
+            //  when `leftProducer.isProducesDifferentValueOverTime()`
+            ArrayList<Task> producerTasks = left.isPresent()
+                ? leftTasks
+                : rightTasks;
+            producerTasks.forEach(visitor::execute);
+        }
+
+        private ArrayList<Task> producerTasksOf(ValueProducer producer) {
+            ArrayList<Task> tasks = new ArrayList<>();
+            producer.visitProducerTasks(tasks::add);
+            return tasks;
+        }
+
     }
 }

--- a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java
+++ b/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java
@@ -47,7 +47,9 @@ public interface ValueSupplier {
         NoProducer NO_PRODUCER = new NoProducer();
         UnknownProducer UNKNOWN_PRODUCER = new UnknownProducer();
 
-        boolean isKnown();
+        default boolean isKnown() {
+            return true;
+        }
 
         boolean isProducesDifferentValueOverTime();
 
@@ -98,10 +100,6 @@ public interface ValueSupplier {
     }
 
     class ExternalValueProducer implements ValueProducer {
-        @Override
-        public boolean isKnown() {
-            return true;
-        }
 
         @Override
         public boolean isProducesDifferentValueOverTime() {
@@ -120,11 +118,6 @@ public interface ValueSupplier {
         public TaskProducer(Task task, boolean content) {
             this.task = task;
             this.content = content;
-        }
-
-        @Override
-        public boolean isKnown() {
-            return true;
         }
 
         @Override
@@ -188,10 +181,6 @@ public interface ValueSupplier {
     }
 
     class NoProducer implements ValueProducer {
-        @Override
-        public boolean isKnown() {
-            return true;
-        }
 
         @Override
         public boolean isProducesDifferentValueOverTime() {

